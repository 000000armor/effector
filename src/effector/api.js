//@flow

// import invariant from 'invariant'
import $$observable from 'symbol-observable'
import warning from 'warning'
import {from, type Stream} from 'most'
import type {Event, Effect, GraphiteMeta} from './index.h'
import * as Kind from '../kind'
import {setProperty} from '../setProperty'
// import {runSyncGraph} from '../../graph'
import * as Cmd from './datatype/cmd'
import * as Step from './datatype/step'
import {walkEvent} from './walk'

export function createEvent<Payload>(name: string): Event<Payload> {
 return eventConstructor({name, domainName: ''})
}

export function createEffect<Payload, Done>(
 name: string,
): Effect<Payload, Done, *> {
 return effectConstructor({name, domainName: ''})
}

function eventConstructor<Payload>({
 name,
 domainName,
}: {
 name: string,
 domainName: string,
}): Event<Payload> {
 const fullName = makeName(name, domainName)

 const cmd: Cmd.Emit = Cmd.emit({
  subtype: 'event',
  fullName,
  runner: createGraphite,
 })
 const step: Step.Single = Step.single(cmd)
 const nextSteps: Step.Multi = Step.multi()
 const stepFull: Step.Seq = Step.seq([step, nextSteps])
 const graphite: GraphiteMeta = {
  cmd,
  step,
  next: nextSteps,
  seq: stepFull,
 }
 const instance = (payload: Payload): Payload =>
  instanceAsEvent.create(payload, fullName)
 function createGraphite(payload: Payload): Payload {
  return instanceAsEvent.create(payload, fullName)
 }
 const instanceAsEvent: Event<Payload> = (instance: any)
 instanceAsEvent.graphite = graphite

 setProperty('create', create, instance)

 setProperty('toString', getType, instance)
 setProperty('getType', getType, instance)
 setProperty('kind', Kind.EVENT, instance)
 setProperty($$observable, () => instance, instance)
 instance.watch = watch
 instance.map = map
 instance.prepend = prepend
 instance.subscribe = subscribe
 instance.to = to
 instance.epic = epic
 // instance.link = (b, ab, ba) => link(instanceAsEvent, b, ab, ba)
 function epic<T>(fn: (Stream<Payload>) => Stream<T>): Event<T> {
  const instance$ = from(instanceAsEvent).multicast()
  const epic$ = fn(instance$).multicast()
  const mapped = eventConstructor({name: `${name}$ ~> *`, domainName})
  epic$.observe(e => {
   mapped.create(e, fullName)
  })
  return mapped
 }
 function to(target, handler?: Function) {
  switch (Kind.readKind(target)) {
   case Kind.STORE:
    return watch(payload => target.setState(payload, handler))
   case Kind.EVENT:
   case Kind.EFFECT:
    return watch(target.create)
   default: {
    throw new TypeError(`Unsupported kind`)
   }
  }
 }
 const seq = 0

 function create(payload, fullName) {
  walkEvent(payload, instanceAsEvent)
  return payload
 }
 function watch(watcher: (payload: Payload, type: string) => any) {
  const runCmd = Step.single(
   Cmd.run({
    runner(newValue: Payload) {
     return watcher(newValue, fullName)
    },
   }),
  )
  instanceAsEvent.graphite.next.data.add(runCmd)
  return () => {
   instanceAsEvent.graphite.next.data.delete(runCmd)
  }
 }

 function subscribe(observer) {
  const unsub = watch(payload => observer.next(payload))
  function unsubscribe() {
   unsub()
  }

  unsubscribe.unsubscribe = unsubscribe

  return unsubscribe
 }
 function map<Next>(fn: Payload => Next) {
  const mapped = eventConstructor({name: `${name} → *`, domainName})
  const computeCmd = Step.single(
   Cmd.compute({
    reduce(_, newValue: Payload, ctx) {
     return fn(newValue)
    },
   }),
  )
  const nextSeq = Step.seq([computeCmd, ...mapped.graphite.seq.data])
  instanceAsEvent.graphite.next.data.add(nextSeq)
  return mapped
 }
 function prepend<Before>(fn: Before => Payload) {
  const contramapped: Event<Before> = eventConstructor({
   name: `* → ${name}`,
   domainName,
  })

  const computeCmd = Step.single(
   Cmd.compute({
    reduce(_, newValue: Before, ctx) {
     return fn(newValue)
    },
   }),
  )
  const nextSeq = Step.seq([computeCmd, ...instanceAsEvent.graphite.seq.data])
  contramapped.graphite.next.data.add(nextSeq)
  return contramapped
 }
 function getType() {
  return fullName
 }
 return (instance: $todo)
}

function effectConstructor<Payload, Done>({
 name,
 domainName,
}: {
 name: string,
 domainName: string,
}): Effect<Payload, Done, *> {
 const instanceAsEvent: Event<Payload> = eventConstructor({
  name,
  domainName,
 })
 const instance: Effect<Payload, Done, any> = (instanceAsEvent: any)
 const eventCreate = instanceAsEvent.create
 const done: Event<{params: Payload, result: Done}> = eventConstructor({
  name: `${name} done`,
  domainName,
 })
 const fail: Event<{params: Payload, error: *}> = eventConstructor({
  name: `${name} fail`,
  domainName,
 })

 // instanceAsEvent.step.data.delete(instanceAsEvent.cmd)
 instance.done = done
 instance.fail = fail
 instance.use = use
 setProperty('create', create, instance)
 setProperty('kind', Kind.EFFECT, instance)
 let thunk = async(value: Payload): Promise<Done> => {
  warning(false, 'no thunk used')
  declare var result: Done
  /*::return result*/
 }
 function use(fn) {
  thunk = fn
 }
 function create(payload: Payload) {
  const throwSymbol = {}
  let syncError = throwSymbol
  let req
  try {
   req = thunk(payload)
  } catch (err) {
   syncError = err
  }
  const isSyncError = syncError !== throwSymbol
  const isPromise = !isSyncError && hasPromise(req)
  //$todo
  const onDone = result => done({params: payload, result})
  const onFail = error => fail({params: payload, error})
  //$todo
  if (isPromise) req.then(onDone, onFail)
  else if (isSyncError) onFail(syncError)
  else onDone(req)
  eventCreate(payload, instanceAsEvent.getType())
  return {
   done() {
    return new Promise(rs => {
     if (isPromise) req.then(result => rs({params: payload, result}))
     else if (!isSyncError) rs({params: payload, result: req})
    })
   },
   fail() {
    return new Promise(rs => {
     if (isPromise) req.catch(error => rs({params: payload, error}))
     else if (isSyncError) rs({params: payload, error: syncError})
    })
   },
   promise() {
    return new Promise((rs, rj) => {
     if (isPromise) req.then(rs, rj)
     else if (isSyncError) rj(syncError)
     else rs(req)
    })
   },
  }
 }

 return instance
}

function hasPromise(req: mixed): boolean %checks {
 return (
  typeof req === 'object' && req !== null && typeof req.then === 'function'
 )
}

function makeName(name, domainName) {
 return [name, domainName].filter(str => str.length > 0).join('/')
}
