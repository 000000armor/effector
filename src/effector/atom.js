//@flow

export class Atom<A> {
 value: A
 get(): A {
  return this.value
 }
 set(_: A) {
  this.value = _
 }
}

export function atom<A>(defaultState: A): Atom<A> {
 const result: Atom<A> = new Atom()
 result.set(defaultState)
 return result
}
