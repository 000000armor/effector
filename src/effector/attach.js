//@flow

import {combine} from './combine'
import {createEffect, onSettled} from './createEffect'
import {applyParentEventHook} from './createUnit'
import {onConfigNesting} from './config'
import {getGraph, getStoreState} from './getter'
import {own} from './own'
import {is} from './is'
import {step} from './typedef'
import {launch} from './kernel'
import {addToReg} from './createNode'

export function attach(config) {
  onConfigNesting(config, (injectedData, userConfig) => {
    config = userConfig
  })
  const {source, effect, mapParams} = config
  const attached = createEffect(config)
  const {runner} = getGraph(attached).scope

  let runnerSteps
  const runnerFn = (
    {params, req},
    {finally: anyway, effect},
    {a: states, page},
  ) =>
    launch({
      target: effect,
      params: {
        params: mapParams(params, states),
        req: {
          rs: onSettled({
            params,
            req,
            ok: true,
            anyway,
            page,
          }),
          rj: onSettled({
            params,
            req,
            ok: false,
            anyway,
            page,
          }),
        },
      },
      page,
      defer: true,
    })

  if (source) {
    let state
    if (is.store(source)) state = source
    else {
      state = combine(source)
      own(attached, [state])
    }
    const readStateRef = step.mov({
      from: 'store',
      store: getStoreState(state),
      to: 'a',
    })
    runnerSteps = [
      /* let another side-effects run first */
      step.run({fn: _ => _}),
      /* read state. assumed it already stable here because of previous step */
      readStateRef,
      /* no need for step.run because of first step */
      step.compute({fn: runnerFn}),
    ]
    addToReg(readStateRef, runner.reg)
  } else {
    runnerSteps = [step.run({fn: runnerFn})]
  }
  runner.scope.effect = effect
  runner.meta.onCopy.push('effect')
  runner.seq.splice(0, 1, ...runnerSteps)
  applyParentEventHook(effect, attached)
  return attached
}
