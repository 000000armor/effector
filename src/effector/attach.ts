import {combine} from './combine'
import {createEffect, onSettled} from './createEffect'
import {applyParentHook} from './createUnit'
import {onConfigNesting} from './config'
import {getGraph, getStoreState} from './getter'
import {own} from './own'
import {is} from './is'
import {step} from './typedef'
import {launch} from './kernel'
import {addToReg} from './createNode'
import {STORE, EFFECT, REG_A} from './tag'

export function attach(config: any) {
  let injected
  onConfigNesting(config, (injectedData, userConfig) => {
    injected = injectedData
    config = userConfig
  })
  let {source, effect, mapParams} = config
  if (!mapParams)
    mapParams = source
      ? (_: any, source: any) => source
      : (params: any) => params
  const attached = createEffect(config, injected)
  const {runner} = getGraph(attached).scope

  let runnerSteps
  const runnerFn = (
    {params, req}: any,
    {finally: anyway, effect}: any,
    stack: any,
  ) => {
    const rj = onSettled({
      params,
      req,
      ok: false,
      anyway,
      stack,
    })
    let computedParams
    try {
      computedParams = mapParams(params, stack.a)
    } catch (err) {
      return rj(err)
    }
    launch({
      target: effect,
      params: {
        params: computedParams,
        req: {
          rs: onSettled({
            params,
            req,
            ok: true,
            anyway,
            stack,
          }),
          rj,
        },
      },
      page: stack.page,
      defer: true,
    })
  }
  if (source) {
    let state
    if (is.store(source)) state = source
    else {
      state = combine(source)
      own(attached, [state])
    }
    const readStateRef = step.mov({
      from: STORE,
      store: getStoreState(state),
      to: REG_A,
    })
    runnerSteps = [
      /* let another side-effects run first */
      step.run({fn: _ => _}),
      /* read state. assumed it already stable here because of previous step */
      readStateRef,
      /* no need for step.run because of first step */
      step.compute({fn: runnerFn}),
    ]
    addToReg(readStateRef, runner.reg)
  } else {
    runnerSteps = [step.run({fn: runnerFn})]
  }
  runner.scope.effect = effect
  runner.meta.onCopy.push(EFFECT)
  runner.seq.splice(0, 1, ...runnerSteps)
  applyParentHook(effect, attached, EFFECT)
  return attached
}
