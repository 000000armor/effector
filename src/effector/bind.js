//@flow

export const bind = (fn: Function, target: any) => fn.bind(null, target)
export const bind2 = (fn: Function, target: any, arg: any) =>
  fn.bind(null, target, arg)
