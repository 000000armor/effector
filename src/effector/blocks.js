import {step} from './typedef'
import {readRef} from './stateRef'

//prettier-ignore
export const filterChanged = step.filter({
  fn: (data, {state}) => (
    data !== undefined &&
    data !== readRef(state)
  ),
})

export const noop = step.compute({
  fn: n => n,
})
