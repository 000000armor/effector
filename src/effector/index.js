//@flow

/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball of wibbly wobbly time-y wimey stuff
 */

export {combine, combine as createStoreObject} from './combine'
export {sample} from './sample'

export type {Domain} from './domain'
export {createDomain} from './domain'

export type {Event} from './event'
export {createEvent, merge, forward, fromObservable} from './event'

export type {Effect} from './effect'
export {createEffect} from './effect'

export type {Store} from './store'
export {
  createStore,
  setStoreName,
  extract,
  createApi,
  restore,
  restoreEvent,
  restoreEffect,
  restoreObject,
  withProps,
} from './store'

export {
  Kind,
  clearNode,
  createGraph as createNode,
  step,
  traverse,
} from './stdlib'
export {launch} from './kernel'
export type {kind} from './stdlib'

export {
  invariant,
  warning,
  isUnit,
  isStore,
  isEvent,
  isEffect,
  isDomain,
  is,
} from './validate'
export {version} from 'effector/flags'

//eslint-disable-next-line
export * as blocks from './blocks'
