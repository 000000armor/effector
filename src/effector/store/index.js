//@flow

export {storeFabric} from './storeFabric'
export {normalizeConfig} from '../config'
export {createStore} from './createStore'
export {setStoreName, storeNaming} from './setStoreName'
export {createStoreObject, extract} from './createStoreObject'
export {createApi} from './createApi'
export {restore, restoreEvent, restoreEffect, restoreObject} from './restore'

export {withProps} from './staticMethods'

export type {Store} from './index.h'
