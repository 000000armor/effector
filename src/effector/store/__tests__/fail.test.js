//@flow

import {createEvent, createStore, is} from 'effector'

test('store.fail is event', () => {
  const store = createStore(0)
  expect(is.event(store.fail)).toBe(true)
})

it('triggers after failed .on', () => {
  const fn = jest.fn()
  const trigger = createEvent()
  const store = createStore(0)
  store.on(trigger, () => {
    throw new Error('Unknown error')
  })
  store.fail.watch(e => fn(e))

  store.setState(1)
  expect(fn).not.toBeCalled()
  trigger()
  expect(fn).toBeCalledTimes(1)
  expect(fn.mock.calls).toMatchSnapshot()
})

it('triggers after failed .map', () => {
  const fn = jest.fn()
  const store = createStore(0)
  const mappedStore = store.map(state => {
    if (state > 5) throw new Error('Unknown error')
    return state
  })
  mappedStore.fail.watch(e => fn(e))

  store.setState(1)
  expect(fn).not.toBeCalled()
  store.setState(6)
  expect(fn).toBeCalledTimes(1)
  expect(fn.mock.calls).toMatchSnapshot()
})
