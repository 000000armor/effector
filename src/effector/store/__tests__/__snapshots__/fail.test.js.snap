// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`triggers after failed .map 1`] = `
Array [
  Array [
    Object {
      "error": [Error: Unknown error],
      "state": 1,
    },
  ],
]
`;

exports[`triggers after failed .on 1`] = `
Array [
  Array [
    Object {
      "error": [Error: Unknown error],
      "state": 1,
    },
  ],
]
`;
